packages/financial-templates-lib/src/price-feed/URTHPriceFeed.ts

```diff
@@ -0,0 +1,192 @@
+  ) {
+    super();
+
+    this.uuid = `Polygon-${index}`;
+
+    this.priceHistory = [];
+
+    this.convertPriceFeedDecimals = (number) => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number.
+      // Note: Must ensure that `number` has no more decimal places than `priceFeedDecimals`.
+      return Web3.utils.toBN(
+        parseFixed(number.toString().substring(0, priceFeedDecimals), priceFeedDecimals).toString()
+      );
+    };
+  }
+  // Updates the internal state of the price feed. Should pull in any async data so the get*Price methods can be called.
+  // Will use the optional ancillary data parameter to customize what kind of data get*Price returns.
+  // Note: derived classes *must* override this method.
+  // Note: Eventually `update` will be removed in favor of folding its logic into `getCurrentPrice`.
+  public async update(ancillaryData?: string): Promise<void> {
+    const currentTime = await this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== null && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "URTHPriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTime + this.minTimeBetweenUpdates - currentTime,
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "URTHPriceFeed",
+      message: "Updating URTHPriceFeed",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime,
+    });
+
+    // 1. Construct URL.
+    // See https://polygon.io/docs/stocks/getting-started.
+    const url = "https://api.polygon.io/v1/open-close/" + this.index + `2023-01-09?adjusted=true&apiKey=${this.apiKey}`;
+
+    // 2. Send request.
+    const historyResponse = await this.networker.getJson(url);
+
+    // 3. Check responses.
+    if (!historyResponse?.data || historyResponse.data.length === 0) {
+      throw new Error(`🚨Could not parse price result from url ${url}: ${JSON.stringify(historyResponse)}`);
+    }
+
+    // 4. Parse results.
+    const newHistoricalPricePeriods = historyResponse.data
+      .map((dailyData: any) => ({
+        date: this._dateTimeToSecond(dailyData.date),
+        openPrice: this.convertPriceFeedDecimals(dailyData.open),
+      }))
+      .sort((a: any, b: any) => {
+        // Sorts the data such that the oldest elements come first.
+        return a.date - b.date;
+      });
+
+    // 5. Store results.
+    this.currentPrice = newHistoricalPricePeriods[newHistoricalPricePeriods.length - 1].openPrice;
+    this.priceHistory = newHistoricalPricePeriods;
+    this.lastUpdateTime = currentTime;
+  }
+
+  public getCurrentPrice(): BN | null {
```

Comment:
You can test that this function is returning the correct Price when the bot is run..